auto-value-gson/src/test/java/com/ryanharter/auto/value/gson/AutoValueGsonExtensionTest.java

```diff
@@ -548,4 +547,125 @@ public void simple() {
       .and()
       .generatesSources(expected);
   }
+
+  @Test public void handlesGenericTypes() {
+    JavaFileObject source1 = JavaFileObjects.forSourceString("test.Foo", ""
+        + "package test;\n"
+        + "import com.google.auto.value.AutoValue;\n"
+        + "import com.google.gson.Gson;\n"
+        + "import com.google.gson.TypeAdapter;\n"
+        + "import com.google.gson.reflect.TypeToken;\n"
+        + "@AutoValue public abstract class Foo<A, B, C> {\n"
+        + "  public static TypeAdapter<Foo> typeAdapter(Gson gson, TypeToken<? extends Foo> typeToken) {\n"
+        + "    return new AutoValue_Foo.GsonTypeAdapter(gson, typeToken);"
+        + "  }\n"
+        + "  public abstract C c();\n"
+        + "  public abstract A a();\n"
+        + "  public abstract B b();\n"
+        + "  public abstract String d();\n"
+        + "}"
+    );
+
+    JavaFileObject expected = JavaFileObjects.forSourceString("test.AutoValue_Test", ""
+        + "package test;\n"
+        + "\n"
+        + "import com.google.gson.Gson;\n"
+        + "import com.google.gson.TypeAdapter;\n"
+        + "import com.google.gson.reflect.TypeToken;\n"
+        + "import com.google.gson.stream.JsonReader;\n"
+        + "import com.google.gson.stream.JsonToken;\n"
+        + "import com.google.gson.stream.JsonWriter;\n"
+        + "import java.io.IOException;\n"
+        + "import java.lang.Object;\n"
+        + "import java.lang.Override;\n"
+        + "import java.lang.String;\n"
+        + "import java.lang.reflect.ParameterizedType;\n"
+        + "import java.lang.reflect.Type;\n"
+        + "\n"
+        + "final class AutoValue_Foo<A, B, C> extends $AutoValue_Foo<A, B, C> {\n"
+        + "  AutoValue_Foo(C c, A a, B b, String d) {\n"
+        + "    super(c, a, b, d);\n"
+        + "  }\n"
+        + "\n"
+        + "  public static final class GsonTypeAdapter extends TypeAdapter<Foo> {\n"
```

Comment:
I think @swankjesse mentioned this in Slack, but this really seems like it wants to be
```java
public static final class GsonTypeAdapter<A, B, C> extends TypeAdapter<Foo<A, B, C>>
```
and having a constructor of
```java
GsonTypeAdapter(Gson gson, TypeAdapter<? extends A> aAdapter, TypeAdapter<? extends B> bAdapter, TypeAdapter<? extends C> cAdapter)
```
This requires also that there's a `TypeAdapterFactory` which can do the ugly reflection on the runtime type for the actual generics.

However, currently you're already forcing the user to create a `TypeToken` for (in this case) `Foo` with all of it's generics filled out for registering with the Gson instance. This means that they're effectively fixed at compile-time and not truly generic like a `TypeAdapterFactory` would be.

(It's a shame I didn't look at this when we were actually in the same room instead of now having to talk about this 1000 miles apart)